### PR TITLE
[Feature] allow specify whether a column use zone map

### LINKED ISSUE #20907: [Feature] allow specify whether a column use zone map
## Feature request

In the current implementation, only key columns or certain types of value columns in a duplicate table will have zone map index created by default. Users cannot create zone map index for primary key tables or varchar value columns.

In this feature, besides the default policy, user can specify a column to using zone map index in duplicate table or primary table.

E.g.
create table with `use_zone_map`
```
CREATE TABLE IF NOT EXISTS tbl1 (
    k1      varchar(20),
    v1      varchar(20),
    v2      varchar(20) use_zone_map
) ENGINE=OLAP
PRIMARY KEY(k1)
DISTRIBUTED BY HASH(k1) BUCKETS 1
properties (
"replication_num" = "1"
);

mysql> desc tbl2;
+-------+-------------+------+-------+---------+--------------+
| Field | Type        | Null | Key   | Default | Extra        |
+-------+-------------+------+-------+---------+--------------+
| k1    | varchar(20) | NO   | true  | NULL    |              |
| v1    | varchar(20) | YES  | false | NULL    |              |
| v2    | varchar(20) | YES  | false | NULL    | USE_ZONE_MAP |
+-------+-------------+------+-------+---------+--------------+
3 rows in set (0.00 sec)
```
And also modify column:
```
alter table tbl2 modify column v2 varchar(20);
mysql> desc tbl2;
+-------+-------------+------+-------+---------+-------+
| Field | Type        | Null | Key   | Default | Extra |
+-------+-------------+------+-------+---------+-------+
| k1    | varchar(20) | NO   | true  | NULL    |       |
| v1    | varchar(20) | YES  | false | NULL    |       |
| v2    | varchar(20) | YES  | false | NULL    |       |
+-------+-------------+------+-------+---------+-------+
3 rows in set (0.00 sec)
mysql> alter table tbl2 modify column v2 varchar(20) use_zone_map;
Query OK, 0 rows affected (0.00 sec)
mysql> desc tbl2;
+-------+-------------+------+-------+---------+--------------+
| Field | Type        | Null | Key   | Default | Extra        |
+-------+-------------+------+-------+---------+--------------+
| k1    | varchar(20) | NO   | true  | NULL    |              |
| v1    | varchar(20) | YES  | false | NULL    |              |
| v2    | varchar(20) | YES  | false | NULL    | USE_ZONE_MAP |
+-------+-------------+------+-------+---------+--------------+
3 rows in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by luohaha**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #20907

## Problem Summary(Required) ：
In the current implementation, only key columns or certain types of value columns in a duplicate table will have zone map index created by default. Users cannot create zone map index for primary key tables or varchar value columns.

In this feature, besides the default policy, user can specify a column to using zone map index in duplicate table or primary table.

In primary table, when rows in segment been deleted, the max/min in the zonemap may become inaccurate, but this will only affect the filtering effect and not the correctness.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/storage/rowset/segment_writer.cpp [hunk 1]
```diff
@@ -146,6 +146,10 @@ Status SegmentWriter::init(const std::vector<uint32_t>& column_indexes, bool has
         // TODO(mofei) refactor it to type specification
         opts.need_zone_map = column.is_key() ||
                              (_tablet_schema->keys_type() == KeysType::DUP_KEYS && is_zone_map_key_type(column.type()));
```
Comment (LiShuMing) [inline be/src/storage/rowset/segment_writer.cpp:148]: Can we use zonemap in the `PRIMARY_KEYS `  key type by default for supported column types like `DUP_KEYS `?

File: be/src/storage/rowset/segment_writer.cpp [hunk 2]
```diff
@@ -146,6 +146,10 @@ Status SegmentWriter::init(const std::vector<uint32_t>& column_indexes, bool has
...
+        if (column.use_zone_map() && (_tablet_schema->keys_type() == KeysType::DUP_KEYS ||
+                                      _tablet_schema->keys_type() == KeysType::PRIMARY_KEYS)) {
+            opts.need_zone_map = true;
```
Comment (LiShuMing) [inline be/src/storage/rowset/segment_writer.cpp:151]: Do we need an guard check to make sure data type is supported for the `use_zone_map` columns?

File: be/src/storage/rowset/segment_writer.cpp (hunk 1 above)
Comment (luohaha) [reply be/src/storage/rowset/segment_writer.cpp:148]: Yes, but this need discuss. Keep the default behavior first.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 3]
```diff
@@ -323,7 +323,7 @@ createTableStatement
      ;
 
 columnDesc
-    : identifier type charsetName? KEY? aggDesc? (NULL | NOT NULL)? (defaultDesc | AUTO_INCREMENT)? comment?
+    : identifier type charsetName? KEY? aggDesc? (NULL | NOT NULL)? (defaultDesc | AUTO_INCREMENT)? USE_ZONE_MAP? comment?
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:326]: The zone map index Syntax should consistent with bloom filter index and bitmap index.

Comment (luohaha) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:326]: `PROPERTIES("zone_map_columns" = "k1,k2");`
`ALTER TABLE table1 SET ("zone_map_columns" = "k1,k2,v1");`
So, like this?

File: be/src/storage/rowset/segment_writer.cpp (hunk 2 above)
Comment (luohaha) [reply be/src/storage/rowset/segment_writer.cpp:151]: Check when create table and alter table.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java [hunk 4]
```diff
@@ -258,6 +259,13 @@ public static void analyze(CreateTableStmt statement, ConnectContext context) {
             if (!columnSet.add(columnDef.getName())) {
                 ErrorReport.reportSemanticException(ErrorCode.ERR_DUP_FIELDNAME, columnDef.getName());
             }
+
+            if (columnDef.useZoneMap()) {
```
Comment (jaogoy) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java:263]: How to unify indices that depending on only one column and on several columns.

**Pull request closed without merging**